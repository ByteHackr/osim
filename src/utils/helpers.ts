import { toRaw, isRef, isReactive, isProxy, ref, toRef, watch } from 'vue';
import * as R from 'ramda';

export function vRef(prop: Record<string, any>, property: string, defaultValue: any) {
  const reffedProp = toRef(prop, property);
  watch(reffedProp, (value) => flexRef.value = value);
  const flexRef = reffedProp.value === undefined ? ref(defaultValue) : reffedProp;
  return flexRef;
}

export function deepToRaw<T extends Record<string, any>>(sourceObj: T): T {
  const objectIterator = (input: any): any => {
    if (Array.isArray(input)) {
      return input.map((item) => objectIterator(item));
    }
    if (isRef(input) || isReactive(input) || isProxy(input)) {
      return objectIterator(toRaw(input));
    }
    if (input && typeof input === 'object') {
      return Object.keys(input).reduce((acc, key) => {
        acc[key as keyof typeof acc] = objectIterator(input[key]);
        return acc;
      }, {} as T);
    }
    return input;
  };

  return objectIterator(sourceObj);
}

export function deepCopyFromRaw<T extends Record<string, any>>(sourceObj: T): T {
  return JSON.parse(JSON.stringify(deepToRaw(sourceObj)));
}

// eslint-disable-next-line max-len
// https://stackoverflow.com/questions/14446511/most-efficient-method-to-groupby-on-an-array-of-objects
export const groupBy = <T>(
  array: T[],
  predicate: (value: T, index: number, array: T[]) => string,
) =>
    array.reduce(
      (acc, value, index, array) => {
        (acc[predicate(value, index, array)] ||= []).push(value);
        return acc;
      },
    {} as { [key: string]: T[] },
    );

export const assignKeyValue = (object: Record<string, any>, key: string, value: any = null) => {
  object[key] = value;
  return object;
};

export const objectMap = (object: Record<string, any>, mapFn: (key: string, value: any) => any) =>
  Object.keys(object).reduce(
    (acc, key) => assignKeyValue(acc, key, mapFn(key, object[key])),
    {} as Record<string, any>,
  );

export const sortedByGroup = <T extends Record<string, any>>(array: T[], key: string) =>
  groupBy(
    array
      .filter((item: T) => item[key])
      .sort((itemA: T, itemB: T) => itemA[key].localeCompare(itemB[key])),
    (item: T) => item[key],
  );

type DeepMappable = any[] | Record<string, any>

const isNonEmptyArray = (value: any) => R.is(Array, value) && value.length > 0;
const isNonArrayObject = (value: any) => R.is(Object, value) && !R.is(Array, value);
const isDeepMappable = (value: DeepMappable) => isNonEmptyArray(value) || isNonArrayObject(value);

export const deepMap = (transform: (arg: any) => any, object: DeepMappable): any =>
  R.map(
    (val: any) => isDeepMappable(val)
      ? deepMap(transform, val)
      : transform(val)
    , object
  );

export const cveRegex = /^\s*(CVE-\d{4}-\d{4,7})\s*$/;
