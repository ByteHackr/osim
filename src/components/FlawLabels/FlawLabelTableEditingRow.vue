<script setup lang="ts">
import { StateEnum } from '@/generated-client';
import type { ZodFlawType } from '@/types';
import { FlawLabelTypeEnum } from '@/types/zodFlaw';
import { watchedRef } from '@/utils/helpers';

import FlawLabelsContributor from './FlawLabelsContributor.vue';

const { contextLabels, initalLabel } = defineProps<{
  contextLabels?: string[];
  initalLabel?: NonNullable<ZodFlawType['labels']>[number];
}>();

const emit = defineEmits<{
  cancel: [];
  save: [label: Omit<NonNullable<ZodFlawType['labels']>[number], 'uuid'>];
}>();

const [labelState, hasStateChanged] = watchedRef<StateEnum>(initalLabel?.state ?? StateEnum.New);
const [labelName, hasNameChanged] = watchedRef(initalLabel?.label ?? '');
const [labelContributor, hasContributorChanged] = watchedRef(initalLabel?.contributor ?? '');

const emitSave = () => {
  // if nothing changed, do not emit save
  if (
    !hasStateChanged.value
    && !hasNameChanged.value
    && !hasContributorChanged.value
  ) {
    return emit('cancel');
  }

  // if label name is empty, do nothing
  if (!labelName.value) {
    return;
  }

  emit('save', {
    ...initalLabel,
    state: labelState.value,
    label: labelName.value,
    contributor: labelContributor.value,
    relevant: initalLabel?.relevant ?? true,
    type: initalLabel?.type ?? FlawLabelTypeEnum.CONTEXT_BASED,
  });
};
</script>

<template>
  <td>
    <select v-model="labelState" class="form-select">
      <option
        v-for="state in StateEnum"
        :key="state"
        :value="state"
      >{{ state }}</option>
    </select>
  </td>
  <td
    :class="{
      'fw-bold': labelState === 'REQ',
      'text-decoration-line-through': !initalLabel?.relevant,
    }"
  >
    <template v-if="initalLabel?.type !== FlawLabelTypeEnum.PRODUCT_FAMILY">
      <select
        v-model="labelName"
        class="form-select"
        required
      >
        <option
          v-for="label in contextLabels"
          :key="label"
          :value="label"
        >{{ label }}</option>
      </select>
    </template>
    <template v-else>
      {{ initalLabel?.label }}
    </template>
  </td>
  <td>
    <FlawLabelsContributor v-model="labelContributor" />
  </td>
  <td>
    <div class="actions">
      <button
        type="button"
        class="btn btn-sm btn-dark"
        title="Save"
        @click="emitSave"
      >
        <i class="bi bi-check2" />
      </button>
      <button
        type="button"
        class="btn btn-sm btn-dark"
        title="Cancel"
        @click="emit('cancel')"
      >
        <i class="bi bi-x" />
      </button>
    </div>
  </td>
</template>

<style scoped lang="scss">
input,
select {
  padding: 0.15rem 0.5rem;
}
</style>
