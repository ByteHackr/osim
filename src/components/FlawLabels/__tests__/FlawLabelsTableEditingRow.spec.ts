import { mountWithConfig } from '@/__tests__/helpers';

import FlawLabelTableEditingRow from '../FlawLabelTableEditingRow.vue';

describe('flawLabelsTableEditingRow', () => {
  it('should render', () => {
    const wrapper = mountWithConfig(FlawLabelTableEditingRow, { shallow: true });

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should not emit save event if nothing changed', async () => {
    const wrapper = mountWithConfig(FlawLabelTableEditingRow);

    await wrapper.find('button[title="Save"]').trigger('click');

    expect(wrapper.emitted()).not.toHaveProperty('save');
    expect(wrapper.emitted()).toHaveProperty('cancel');
  });

  it('should emit save event if value changed', async () => {
    const wrapper = mountWithConfig(FlawLabelTableEditingRow, {
      props: {
        contextLabels: ['test'],
      },
    });

    await wrapper.findAll('select')[1].setValue('test');
    await wrapper.find('button[title="Save"]').trigger('click');

    expect(wrapper.emitted()).toHaveProperty('save');
    expect(wrapper.emitted('save')).toEqual([[expect.objectContaining({ label: 'test' })]]);
  });
});
