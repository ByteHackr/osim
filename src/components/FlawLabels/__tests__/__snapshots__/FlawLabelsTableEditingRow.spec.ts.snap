// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`flawLabelsTableEditingRow > should render 1`] = `
"<td data-v-627548cb=""><select data-v-627548cb="" class="form-select">
    <option data-v-627548cb="" value="NEW">NEW</option>
    <option data-v-627548cb="" value="REQ">REQ</option>
    <option data-v-627548cb="" value="SKIP">SKIP</option>
    <option data-v-627548cb="" value="DONE">DONE</option>
  </select></td>
<td data-v-627548cb="" class="text-decoration-line-through"><select data-v-627548cb="" class="form-select" required=""></select></td>
<td data-v-627548cb="">
  <flaw-labels-contributor-stub data-v-627548cb="" modelvalue=""></flaw-labels-contributor-stub>
</td>
<td data-v-627548cb="">
  <div data-v-627548cb="" class="actions"><button data-v-627548cb="" type="button" class="btn btn-sm btn-dark" title="Save"><i data-v-627548cb="" class="bi bi-check2"></i></button><button data-v-627548cb="" type="button" class="btn btn-sm btn-dark" title="Cancel"><i data-v-627548cb="" class="bi bi-x"></i></button></div>
</td>"
`;
